aten/src/ATen/ZeroTensorFallback.cpp

```diff
@@ -9,7 +9,77 @@
 
 namespace at {
 
-  // TODO: add a note explaining the design decisions
+  /*
+  * The goal of ZeroTensors is to replace the undefined tensors used in the forward
+  * and backward mode AD which are currently materialized to a tensor full of zeros
+  * with a new type of zero tensor that use O(1) memory (tensors with a storage
+  * pointing to nullptr). Having undefined tensors is a common occurrence in the
+  * forward mode AD. Currently these undefined tensors are materialized to zero
+  * tensors. Not only the materialization is wasteful, but the cases where
+  * computationally heavy operations like matrix multiplies are called for these
+  * tensors along with some non-zero tensors provide opportunities for perf
+  * improvement. With zero tensors, expensive operations like matmul on zero
+  * tensors will become O(1).
+  *
+  * Note that as of writing this, ZeroTensors do not propagate nans or other
+  * extreme values. This is rooted in how the kernels are implemented for
+  * ZeroTensors. For example, ZT * (non-ZT) = ZT, i.e., we don't check whether the
+  * non-ZT has nans for example.
+  *
+  * Design:
+  *
+  * 1. ZeroTensors are regular tensors with TensorOptions (as usual), and a storage
+  *    pointing to nullptr (aliasing semantic is correctly maintained), and a
+  *    ZeroTensor dispatch key set.
+  *
+  * 2. Zero tensors are immutable.
+  *     - As a result, in-place operations on these tensors will be disallowed. This is
+  *       done to prevent data race in the case of multithreading (when two threads try
+  *       to read the same zero tensor and materialize it in-place).
+  *     - Zero tensors shouldn’t be passed as an output tensor in out= operations.
```

Comment:
Comment about out= should already be implied.
Just keep the "Zero tensors are immutable" and under that "This is done to prevent data race in the case of multithreading (when two threads try to read the same zero tensor and materialize it in-place)."